torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -1197,6 +1197,113 @@ class TensorExprFuser {
   }
 #undef REQ
 
+  // Move the given user of `aten::cat` op to its inputs.
+  Node* moveCatAfterUse(
+      Node* cat,
+      Node* user,
+      std::shared_ptr<Graph> subgraph) {
+    // Example IR:
+    //   %1 = ...
+    //   %2 = ...
+    //   %3 = prim::ListConstruct(%1, %2)
+    //   %4 = aten::cat(%3, ...)
+    //   %5 = aten::relu(%4)
+    //   return (%5)
+    //
+    // To be transformed to:
+    //   %1 = ...
+    //   %2 = ...
+    //   %5.1 = aten::relu(%1)
+    //   %5.2 = aten::relu(%2)
+    //   %3 = prim::ListConstruct(%5.1, %5.2)
+    //   %4 = aten::cat(%3, ...)
+    //   return (%4)
+
+    TORCH_INTERNAL_ASSERT(cat->output()->hasUses());
+    TORCH_INTERNAL_ASSERT(cat->output()->uses().size() == 1);
+    TORCH_INTERNAL_ASSERT(cat->input(0)->node()->kind() == prim::ListConstruct);
+    auto cat_list = cat->input(0)->node();
+    auto cat_inputs = cat_list->inputs();
+
+    std::unordered_map<Value*, Value*> new_cat_inputs;
+    for (auto inp : cat_inputs) {
+      auto new_cat_input = subgraph->createClone(
+          user, [&](Value* k) { return (k == cat->output()) ? inp : k; });
+      new_cat_input->output()->setType(inp->type());
```

Comment:
is it possible we want the dtype of the output of cat here ? In case we have something like `sigmoid` which has type promotion ?